dyn_sim/sys/planar/segway.py

```diff
@@ -1 +1,212 @@
-# [TODO] write this
+import numpy as np
+from matplotlib.axes import Axes
+
+from dyn_sim.sys.dyn_sys import CtrlAffineSystem
+from dyn_sim.util.sim_utils import draw_circle
+
+# constants
+g = 9.80665  # gravitational acceleration
+
+
+class Segway(CtrlAffineSystem):
+    """Segway object.
+
+    State vector is in R^4.
+        state vector s is in order [p, phi, dp, dphi]
+        p - horizontal position, positive right
+        phi - rotational position, positive cw
+        dp - time derivative of horizontal position
+        dphi - time derivative of rotational position
+    Input vector is in R.
+        u - motor voltage
+    """
+
+    def __init__(
+        self,
+        m0: float = 52.71,
+        L: float = 0.169,
+        J0: float = 5.108,
+        Km: float = 2 * 1.262,
+        R: float = 0.195,
+        bt: float = 2 * 1.225,
+        l: float = 0.75,
+        mass: float = 44.798,
+    ) -> None:
+        """Initialize a segway.
+
+        Parameters
+        ----------
+        m0 : float
+            Lumped mass of the segway.
+        L : float
+            Length between the center of rotation and center of gravity.
+        J0 : float
+            Lumped inertia of the segway.
+        Km : float
+            Motor Torque constant.
+        R : float
+            Wheel radius.
+        bt : float
+            Motor damping constant.
+        l : float
+            Length between center of rotation and tip of segway arm.
+        mass : float
+            Mass of upper segway arm (name spelled out to prevent collision).
+        """
+        super(Segway, self).__init__(4, 1, False)
+        self._m0 = m0
+        self._L = L
+        self._J0 = J0
+        self._Km = Km
+        self._R = R
+        self._bt = bt
+        self._l = l
+        self._mass = mass
+
+    def fdyn(self, t: float, s: np.ndarray) -> np.ndarray:
+        """Segway autonomous dynamics.
+
+        Parameters
+        ----------
+        t : float
+            Time. Unused, included for API compliance.
+        s : np.ndarray, shape=(4,)
+            State of segway.
+
+        Returns
+        -------
+        _fdyn : np.ndarray, shape=(4,)
+            Time derivatives of states from autonomous dynamics.
+        """
+        assert s.shape == (4,)
+
+        # states
+        _, phi, dp, dphi = s
+        sinphi = np.sin(phi)
+        cosphi = np.cos(phi)
+
+        # inertia matrix
+        D = np.array(
+            [
+                [self._m0, self._m * self._L * cosphi],
+                [self._m * self._L * cosphi, self._J0],
+            ]
+        )
+        Dinv = np.linalg.inv(D)
```

Comment:
i think you should analytically derive Dinv and define it instead of doing repeated inv calls, which are pretty expensive when evaluating the dynamics in an ODE solver